File: algorithms/sorting/shell-sort/go/ShellSort.go
package main

import "fmt"

func ShellSort(arr []int) []int {
	n := len(arr)
	for gap := n / 2; gap > 0; gap /= 2 {
		for i := gap; i < n; i++ {
			temp := arr[i]
			j := i
			for j >= gap && arr[j-gap] > temp {
				arr[j] = arr[j-gap]
				j -= gap
			}
			arr[j] = temp
		}
	}
	return arr
}

func main() {
	arr := []int{5, 3, 8, 1, 2, -3, 0}
	fmt.Println(ShellSort(arr))
}

File: algorithms/sorting/shell-sort/rust/shell_sort.rs
pub fn shell_sort(arr: &mut [i32]) {
    let n = arr.len();
    let mut gap = n / 2;
    
    while gap > 0 {
        for i in gap..n {
            let temp = arr[i];
            let mut j = i;
            while j >= gap && arr[j - gap] > temp {
                arr[j] = arr[j - gap];
                j -= gap;
            }
            arr[j] = temp;
        }
        gap /= 2;
    }
}
